mir_build: Integrate "simplification" steps into match-pair-tree creation

Zalathar opened the pull request:
The “simplification” step helps to prepare THIR patterns for lowering into MIR, and originally dates back to the earliest days of MIR in the compiler.

Over time, various intermediate data structures have been introduced (e.g. `MatchPair`, later renamed to `MatchPairTree`) that reduce the need for a separate simplification step, because some of the necessary simplifications can be built into the construction of those intermediate structures instead. This PR continues that process to its logical conclusion and removes the simplification step entirely, by integrating its remaining responsibilities into match-pair-tree creation: flattening “irrefutable” nodes, collecting bindings/ascriptions in flat lists, and sorting or-patterns after other subpatterns.

This has a few immediate benefits:
- We can remove `TestCase::Irrefutable`, which was not allowed to exist after simplification, and was much larger than other test-case variants.
- We can make `MatchPairTree::place` non-optional, because only irrefutable nodes could fail to have a place.

In the future, this should also help with some ideas I have for simplifying how `AscribeUserType` and `ExpandedConstant` nodes are handled, by representing them as side-data keyed by THIR pattern ID, so that they are no longer their own kinds of THIR pattern node.

Zalathar commented:
cc @Nadrieril

(Looks like I accidentally broke triagebot pings for `rustc_mir_build/src/builder`; there's a fix at #137876.)

Zalathar commented:
Actually, this seems more appropriate:

r? Nadrieril

Nadrieril commented on compiler/rustc_mir_build/src/builder/matches/mod.rs:
```diff
@@ -1271,6 +1270,8 @@ pub(crate) struct MatchPairTree<'tcx> {
     /// parent has succeeded. For example, the pattern `Some(3)` might have an
     /// outer match pair that tests for the variant `Some`, and then a subpair
     /// that tests its field for the value `3`.
+    ///
+    /// Invariant: Or-patterns must be sorted to the end.
     subpairs: Vec<Self>,
```
Why the invariant on subpairs? I don't think we ever rely on it since we need to re-sort the whole candidate whenever we add new match paris to it.

Nadrieril replied:
I would prefer to remove this: the invariant is relevant for `Candidate`s only; adding it here is potentially confusing. We could even remove it from `FlatPat` if we make sure to sort in the `FlatPat` -> `Candidate` conversion step.

Nadrieril commented on compiler/rustc_mir_build/src/builder/matches/match_pair.rs:
```diff
@@ -333,3 +334,9 @@ impl<'tcx> MatchPairTree<'tcx> {
     }
 }
+
+/// Sorts or-patterns to the end of the list, as required by [`FlatPat`] and
+/// its enclosed [`MatchPairTree`] nodes.
+pub(super) fn sort_match_pairs(match_pairs: &mut [MatchPairTree<'_>]) {
+    match_pairs.sort_by_key(|pair| matches!(pair.test_case, TestCase::Or { .. }));
+}
```
While you're at it, could you also replace the `sort_by_key` call here: https://github.com/rust-lang/rust/blob/4559163ccb500affc424fb9228dae5003672ffc7/compiler/rustc_mir_build/src/builder/matches/test.rs?plain=1#L772

Nadrieril reviewed (changes requested):
I like this, anything that removes some complexity in matchy lowering is a success. I left two requests for changes.

Zalathar commented on compiler/rustc_mir_build/src/builder/matches/mod.rs:
```diff
@@ -1271,6 +1270,8 @@ pub(crate) struct MatchPairTree<'tcx> {
     /// parent has succeeded. For example, the pattern `Some(3)` might have an
     /// outer match pair that tests for the variant `Some`, and then a subpair
     /// that tests its field for the value `3`.
+    ///
+    /// Invariant: Or-patterns must be sorted to the end.
     subpairs: Vec<Self>,
```
Yeah, the only reason it's an invariant here is that `Candidate` relies on it, but I suppose it's possible to do a (recursive) sort in `Candidate::from_flat_pat` instead.

Zalathar committed:
Make `MatchPairTree::for_pattern` push its output node to a vector

Zalathar committed:
Make `field_match_pairs` push its output nodes to a vector

Zalathar committed:
Remove `TestCase::Irrefutable`

Zalathar committed:
Populate pattern bindings/ascriptions while building `MatchPairTree`

Zalathar committed:
Remove the separate simplify step for match-pair trees

What remained of this simplification process has been integrated into
construction of the match-pair trees.

Zalathar committed:
Make `MatchPairTree::place` non-optional

As the invariant indicated, this place could only be none for
`TestCase::Irrefutable` nodes, which no longer exist.

Nadrieril commented:
@bors r+
